From 857d45e622eb9ee371e5e2e6d2629956087c9820 Mon Sep 17 00:00:00 2001
From: Neo Fang <neo.fang@intel.com>
Date: Thu, 14 Sep 2023 07:44:22 +0000
Subject: [PATCH] Introduce v4l2loopback module as part of ivi image

Issues: No V4L2loopback driver and device node when system
boots up.

RootCause: V4L2loopback is not enabled in celadon ivi kernel.

Fix Details: Introduce V4L2loopback driver module as part of
celadon kernel, create device node /dev/video50 when system
boots up.

Tracked-On: OAM-112223
Signed-off-by: Neo Fang <neo.fang@intel.com>
---
 groups/kernel/AndroidBoard.mk                  |  2 +-
 .../linux-intel-lts2021/x86_64_defconfig       |  1 +
 groups/v4l2loopback/default                    |  1 +
 groups/v4l2loopback/doc.spec                   | 18 ++++++++++++++++++
 groups/v4l2loopback/false/empty_dir            |  0
 groups/v4l2loopback/true/init.rc               |  2 ++
 6 files changed, 23 insertions(+), 1 deletion(-)
 create mode 120000 groups/v4l2loopback/default
 create mode 100644 groups/v4l2loopback/doc.spec
 create mode 100644 groups/v4l2loopback/false/empty_dir
 create mode 100644 groups/v4l2loopback/true/init.rc

diff --git a/groups/kernel/AndroidBoard.mk b/groups/kernel/AndroidBoard.mk
index e3965d9..75fc21e 100644
--- a/groups/kernel/AndroidBoard.mk
+++ b/groups/kernel/AndroidBoard.mk
@@ -223,7 +223,7 @@ $(LOCAL_KERNEL_PATH)/copy_modules: $(LOCAL_KERNEL)
 {{#camera_cos_hack}}
 ifeq ($(KERNEL_MODULES_ROOT),vendor/lib/modules)
 	$(hide) mkdir -p $(PRODUCT_OUT)/root/vendor/lib/modules/
-	$(hide) for f in atomisp-css2401a0_v21.ko videobuf-core.ko videobuf-vmalloc.ko; do \
+	$(hide) for f in atomisp-css2401a0_v21.ko videobuf-core.ko videobuf-vmalloc.ko v4l2loopback.ko; do \
 		find $(LOCAL_KERNEL_PATH)/lib/modules/ -name $$f -exec cp {} $(PRODUCT_OUT)/root/vendor/lib/modules/ \; ;\
 		done
 
diff --git a/groups/kernel/gmin64/config-lts/linux-intel-lts2021/x86_64_defconfig b/groups/kernel/gmin64/config-lts/linux-intel-lts2021/x86_64_defconfig
index 7b4928e..d026c7f 100644
--- a/groups/kernel/gmin64/config-lts/linux-intel-lts2021/x86_64_defconfig
+++ b/groups/kernel/gmin64/config-lts/linux-intel-lts2021/x86_64_defconfig
@@ -3776,6 +3776,7 @@ CONFIG_VIDEO_V4L2_SUBDEV_API=y
 # CONFIG_VIDEO_ADV_DEBUG is not set
 # CONFIG_VIDEO_FIXED_MINOR_RANGES is not set
 # end of Video4Linux options
+CONFIG_VIDEO_V4L2LOOPBACK=m
 
 #
 # Media controller options
diff --git a/groups/v4l2loopback/default b/groups/v4l2loopback/default
new file mode 120000
index 0000000..02e4a84
--- /dev/null
+++ b/groups/v4l2loopback/default
@@ -0,0 +1 @@
+false
\ No newline at end of file
diff --git a/groups/v4l2loopback/doc.spec b/groups/v4l2loopback/doc.spec
new file mode 100644
index 0000000..68e68ed
--- /dev/null
+++ b/groups/v4l2loopback/doc.spec
@@ -0,0 +1,18 @@
+=== Overview
+
+Loading v4l2loopback Driver.
+
+--- deps
+
+    - sepolicy
+
+==== Options
+
+--- true
+Will enable v4l2loopback
+
+--- false
+empty dir.
+
+--- default
+when not explicitly selected in mixin spec file, the default option will be used.
diff --git a/groups/v4l2loopback/false/empty_dir b/groups/v4l2loopback/false/empty_dir
new file mode 100644
index 0000000..e69de29
diff --git a/groups/v4l2loopback/true/init.rc b/groups/v4l2loopback/true/init.rc
new file mode 100644
index 0000000..06efd75
--- /dev/null
+++ b/groups/v4l2loopback/true/init.rc
@@ -0,0 +1,2 @@
+on post-fs
+    insmod /vendor/lib/modules/v4l2loopback.ko devices=1 video_nr=50 card_label="LicCam" exclusive_caps=1 broadcast=1 nlgroup=22
-- 
2.17.1

